### ISSUE TITLE
Refactor draw list and remove m_nextToDraw

### ISSUE THREAD

**Issue by KubaO** [wxMaxima-developers/wxmaxima#1445]
The `m_nextToDraw` should be removed.

The draw list implements a depth-first node tree traversal, using the `m_nextToDraw` pointer. That is redundant, since each cell knows what are the cells on its draw list, and can guide iteration programmatically, without having to store a pointer.

I imagine that the pointer came to be to make the depth first traversal possible without maintaining a node stack during iteration, but:

1. Instead of a relatively small node stack in the iterator - a few dozen pointers deep in the most contrived benchmarks, and normally just a few pointers deep, there's the memory cost of a pointer in all nodes.

2. The code that manages the draw list is brittle, even if it's well factored out (`SetNextToDraw` is called in just a couple of places).

3. The `m_nextToDraw` pointer is the last major user of `CellPtr`, and many cells end up having to instantiate the cell pointer control block. The goal of `CellPtr` is to track just a few cells for selection and similar user-interaction purposes. Its use in the draw list was a necessary evil and was meant to be temporary.

The bigger problem is that the draw list isn't uniformly used. There are some cell types whose `Draw` method draws the inner cells that should be drawn. In other cases, they are drawn while the draw list is iterated. This should be investigated and unified.

**Comment by gunterkoenigsmann** [comment]
That double list approach odd hard to describe, hard to understand and error-prone.
But it never occurred to me that it might be unnecessary, as well...

**Comment by KubaO** [comment]
Yep - there are bits of code all over the place that dealt with draw list pointers. When removed, all cells shrink by 8 bytes, and the CellPtr control block overhead also goes away - that's one less allocation per cell. It will not be a huge speed improvement, but a small incremental one. Getting rid of it blocks other improvements, such as the `CellList` refactoring that gets rid of the remaining cell pointers.

**Comment by KubaO** [comment]
@gunterkoenigsmann I'd like you to test the https://github.com/wxMaxima-developers/wxmaxima/tree/feature/KubaO/remove-nexttodraw branch if you can. The thing to look for is any artifacts related to draw list - the `m_nextToDraw` pointer has been completely removed. I'm otherwise ready to merge it.

**Comment by KubaO** [comment]
Interestingly enough, this change slashes the performance benchmark test time from ~11 to ~6 seconds on Bionic.

On Windows, using MSVC 2019, the `hs(17)` benchmark got ~8x faster: time went down from ~4min to ~30s.

```
hs(n):=ratsimp(ratexpand(product(x-a[i],i,1,n)));
hs(17);
```

Alas, on Focal, the benchmark time seems unaffected. It'd have to be checked into - it's a clear pattern now, there's some extraneous overhead on Focal that doesn't seem to have much to do with performance of our code. In spite of improvements on other platforms, the timing of the performance test on Focal (in the test suite) has been stuck around 20s for a good while now.

**Comment by gunterkoenigsmann** [comment]
That your work makes the code simpler and therefore is very important was obvious. But that it makes the code so much faster again caught me by surprise.

Unfortunately I seem to be unable to compile it, currently:

{{{{
[ 56%] Building CXX object src/CMakeFiles/wxmaxima.dir/MathParser.cpp.o
In file included from /home/gunter/src/wxmaxima/src/cells/FracCell.h:34,
                 from /home/gunter/src/wxmaxima/src/MathParser.h:42,
                 from /home/gunter/src/wxmaxima/src/MathParser.cpp:33:
/home/gunter/src/wxmaxima/src/cells/ParenCell.h:79:8: error: ‘void ParenCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   79 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/ParenCell.h:80:9: error: ‘Cell* ParenCell::GetNextToDraw() const’ marked ‘override’, but does not override
   80 |   Cell *GetNextToDraw() const override { return m_nextToDraw; }
      |         ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.h:42,
                 from /home/gunter/src/wxmaxima/src/MathParser.cpp:33:
/home/gunter/src/wxmaxima/src/cells/FracCell.h:86:8: error: ‘void FracCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   86 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/FracCell.h:87:9: error: ‘Cell* FracCell::GetNextToDraw() const’ marked ‘override’, but does not override
   87 |   Cell *GetNextToDraw() const override { return m_nextToDraw; }
      |         ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.cpp:38:
/home/gunter/src/wxmaxima/src/cells/SubCell.h:54:8: error: ‘void SubCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   54 |   void SetNextToDraw(Cell *next) override { m_nextToDraw = next; }
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/SubCell.h:55:9: error: ‘Cell* SubCell::GetNextToDraw() const’ marked ‘override’, but does not override
   55 |   Cell *GetNextToDraw() const override { return m_nextToDraw; }
      |         ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.cpp:39:
/home/gunter/src/wxmaxima/src/cells/SqrtCell.h:72:8: error: ‘void SqrtCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   72 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/SqrtCell.h:74:9: error: ‘Cell* SqrtCell::GetNextToDraw() const’ marked ‘override’, but does not override
   74 |   Cell *GetNextToDraw() const override { return m_nextToDraw; }
      |         ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.cpp:40:
/home/gunter/src/wxmaxima/src/cells/LimitCell.h:66:8: error: ‘void LimitCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   66 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/LimitCell.h:67:9: error: ‘Cell* LimitCell::GetNextToDraw() const’ marked ‘override’, but does not override
   67 |   Cell *GetNextToDraw() const override {return m_nextToDraw;}
      |         ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.cpp:41:
/home/gunter/src/wxmaxima/src/cells/ListCell.h:73:8: error: ‘void ListCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   73 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/ListCell.h:74:9: error: ‘Cell* ListCell::GetNextToDraw() const’ marked ‘override’, but does not override
   74 |   Cell *GetNextToDraw() const override { return m_nextToDraw; }
      |         ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.cpp:48:
/home/gunter/src/wxmaxima/src/cells/SumCell.h:73:9: error: ‘Cell* SumCell::GetNextToDraw() const’ marked ‘override’, but does not override
   73 |   Cell *GetNextToDraw() const override { return m_nextToDraw; }
      |         ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/SumCell.h:75:8: error: ‘void SumCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   75 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.cpp:49:
/home/gunter/src/wxmaxima/src/cells/IntCell.h:75:8: error: ‘void IntCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   75 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/IntCell.h:77:9: error: ‘Cell* IntCell::GetNextToDraw() const’ marked ‘override’, but does not override
   77 |   Cell *GetNextToDraw() const override { return m_nextToDraw; }
      |         ^~~~~~~~~~~~~
In file included from /home/gunter/src/wxmaxima/src/MathParser.cpp:50:
/home/gunter/src/wxmaxima/src/cells/FunCell.h:80:8: error: ‘void FunCell::SetNextToDraw(Cell*)’ marked ‘override’, but does not override
   80 |   void SetNextToDraw(Cell *next) override;
      |        ^~~~~~~~~~~~~
/home/gunter/src/wxmaxima/src/cells/FunCell.h:81:9: error: ‘Cell* FunCell::GetNextToDraw() const’ marked ‘override’, but does not override
   81 |   Cell *GetNextToDraw() const override {return m_nextToDraw;}
      |         ^~~~~~~~~~~~~
make[2]: *** [src/CMakeFiles/wxmaxima.dir/build.make:381: src/CMakeFiles/wxmaxima.dir/MathParser.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:5766: src/CMakeFiles/wxmaxima.dir/all] Error 2
make: *** [Makefile:182: all] Error 2
}}}}
On Groovy wxMaxima interprets about 20 maxima commands per second, currently. But that is way less than on windows, too. My guess was that wxMaxima is fast enough to start lengthy processes like recalculating the worksheet the moment maxima requests more data in. But as that is a guess I might be entirely wrong, this time, again.

**Comment by gunterkoenigsmann** [comment]
Now it compiles. What still is broken (hope it worked before the refactoring):

-  A `diff` cell that is long enough to be broken into lines doesn't end with the 2nd and 3rd argument of `diff`, but with a 2d fraction that shows the diff variable. Don't know if that is a feature or a bug, though.
- I cannot select the entire diff, if it is broken into lines.
- If the demoninator and/or numerator of a fraction is long enough to be broken into lines the fraction is broken into lines, but the denominator isn't.
- Trying to select the entire denominator or numerator in this case results in triggering an assert about a missing cell this cell belongs to.
- Parenthesis cells are broken into lines, if they are too long for the current line, but if they contain parenthesis, that are too long for the current line, the inner parenthesis isn't broken into lines. Trying to select that inner cell again results in a "claims to have no group Cell it belongs to".

This seem to be the general error types I see. I guess it makes no sense to repeat my observations for integral() cells and the other cells that can be found in the manual testsuite.

Us no more breaking the inner cells into lines might be the cause of at least part of the big speedups this change made. But I hope that this assumption of mine isn't true ;-(

**Comment by KubaO** [comment]
The benchmark that tests this does break everything up, it’s just one formula that turns into a very long document that way.

Thank you for spotting those problems. I will investigate how the new and old iteration behaviors differ and fix the corner cases using those cases you identified.

There’s also a now-useless concept of inner cells, because they really not need to be touched all over the place unless they are visible and then they appear on the draw list. So most likely as I continue with this pursuit I’ll remove inner cell list and make draw list take over that job.

**Comment by gunterkoenigsmann** [comment]
>  The benchmark that tests this does break everything up, it’s just one formula that turns into a very long document that way.

That is correct. But layouting equations always is suspiciously fast if the mechanism that breaks  long parenthesis, fractions or other objects into lines even if they are inside other parenthesis, fractions or other objects that have to be broken into lines to make that possible, as well. Not that fixing breaking nested objects into lines undoes most of this speed-up...

> There’s also a now-useless concept of inner cells, because they really not need to be touched all over the place unless they are visible and then they appear on the draw list.

I am already looking forward to finding out how the new code will look like.

**Comment by KubaO** [comment]
Wrong issue # in dc2fc70 :)
